dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/message/ProgramMessageOperationParamMapper.java

```diff
@@ -59,14 +61,34 @@ public class ProgramMessageOperationParamMapper {
   @Transactional(readOnly = true)
   public ProgramMessageQueryParams map(ProgramMessageOperationParams operationParams)
       throws NotFoundException {
+    // TODO(DHIS2-19702): Refactor this method
+    Optional<TrackerEvent> trackerEvent = Optional.empty();
+    Optional<SingleEvent> singleEvent = Optional.empty();
     Enrollment enrollment = getEntity(operationParams.getEnrollment(), Enrollment.class);
-    TrackerEvent event = getEntity(operationParams.getEvent(), TrackerEvent.class);
+    if (operationParams.getEvent() != null) {
+      trackerEvent = findEntity(operationParams.getEvent(), TrackerEvent.class);
+      if (trackerEvent.isEmpty()) {
```

Comment:
why do we throw if we cannot find a trackerEvent, could it not still eb a singleEvent?

Reply:
Yes, but no.
This is the most confusing PR, from the next one things should start to make sense.
At this moment we have `TrackerEvent` and `SingleEvent` hibernate objects that point to the same table `event`, so if we try to get a `TrackerEvent` with the manager and it is not there, we know that there is no event with such a uid in table `event`.
Then based on the type of the program we decide to retrieve it as a single event or not.